### PR TITLE
Add table aws_elastic_beanstalk_application. Closes #151

### LINKED ISSUE #151: Add table aws_elastic_beanstalk_application
**References**
Add any related links that will help us understand the resource, including vendor documentation, related GitHub issues, and Go SDK documentation.
https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.37.26/service/elasticbeanstalk/elasticbeanstalkiface

### PR DESCRIPTION
**Pull request by anisadas**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/aws_elastic_beanstalk_application []

PRETEST: tests/aws_elastic_beanstalk_application

TEST: tests/aws_elastic_beanstalk_application
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_iam_service_linked_role.elasticbeanstalk: Creating...
aws_iam_service_linked_role.elasticbeanstalk: Creation complete after 3s [id=arn:aws:iam::986325076436:role/aws-service-role/elasticbeanstalk.amazonaws.com/AWSServiceRoleForElasticBeanstalk]
aws_elastic_beanstalk_application.named_test_resource: Creating...
aws_elastic_beanstalk_application.named_test_resource: Creation complete after 7s [id=turbottest69498]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

account_id = 986325076436
aws_partition = aws
aws_region = us-east-1
resource_aka = arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbottest69498
resource_name = turbottest69498

Running SQL query: test-get-query.sql
[
  {
    "arn": "arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbottest69498",
    "name": "turbottest69498",
    "tags": {
      "name": "turbottest69498"
    }
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "arn": "arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbottest69498",
    "name": "turbottest69498",
    "tags_src": [
      {
        "Key": "name",
        "Value": "turbottest69498"
      }
    ]
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "arn": "arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbottest69498",
    "name": "turbottest69498",
    "partition": "aws",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest69498"
    },
    "title": "turbottest69498"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbottest69498"
    ],
    "name": "turbottest69498",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest69498"
    },
    "title": "turbottest69498"
  }
]
✔ PASSED

POSTTEST: tests/aws_elastic_beanstalk_application

TEARDOWN: tests/aws_elastic_beanstalk_application

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
select
  name,
  arn,
  description,
  date_created,
  date_updated,
  versions
from
  aws_elastic_beanstalk_application;

+------------------+------------------------------------------------------------------------------+------------------------------------------------------+---------------------+---------------
| name             | arn                                                                          | description                                          | date_created        | date_updated  
+------------------+------------------------------------------------------------------------------+------------------------------------------------------+---------------------+---------------
| turbot           | arn:aws:elasticbeanstalk:us-east-1:986325076436:application/turbot           | <null>                                               | 2020-09-16 10:39:55 | 2020-09-16 10:
| test-anisa       | arn:aws:elasticbeanstalk:us-east-2:986325076436:application/test-anisa       | This application is used to test the steampipe query | 2021-03-16 07:01:27 | 2021-03-16 07:
|                  |                                                                              |                                                      |                     |               
| rups             | arn:aws:elasticbeanstalk:us-east-2:986325076436:application/rups             | <null>                                               | 2020-10-14 11:05:48 | 2020-10-14 11:
| test-application | arn:aws:elasticbeanstalk:us-east-2:986325076436:application/test-application | <null>                                               | 2021-03-12 08:03:08 | 2021-03-12 08:
+------------------+------------------------------------------------------------------------------+------------------------------------------------------+---------------------+---------------

select
  name,
  resource_lifecycle_config ->> 'ServiceRole' as role,
  resource_lifecycle_config -> 'VersionLifecycleConfig' ->> 'MaxAgeRule' as max_age_rule,
  resource_lifecycle_config -> 'VersionLifecycleConfig' ->> 'MaxCountRule' as max_count_rule
from
  aws_elastic_beanstalk_application;

+--------+--------+----------------------------------------------------------------------+------------------------------------------------------------------+
| name   | role   | max_age_rule                                                         | max_count_rule                                                   |
+--------+--------+----------------------------------------------------------------------+------------------------------------------------------------------+
| turbot | <null> | {"Enabled": false, "MaxAgeInDays": 180, "DeleteSourceFromS3": false} | {"Enabled": false, "MaxCount": 200, "DeleteSourceFromS3": false} |
| rups   | <null> | {"Enabled": false, "MaxAgeInDays": 180, "DeleteSourceFromS3": false} | {"Enabled": false, "MaxCount": 200, "DeleteSourceFromS3": false} |
+--------+--------+----------------------------------------------------------------------+------------------------------------------------------------------+
```
</details>

### CODE REVIEW

File: aws/table_aws_elastic_beanstalk_application.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+			{
+				Name:        "description",
+				Description: "User-defined description of the application.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "configuration_Templates",
+				Description: "The names of the configuration templates associated with this application.",
+				Type:        proto.ColumnType_JSON,
+			},
```
Comment (Paulami30) [inline aws/table_aws_elastic_beanstalk_application.go:49]: All the json fields at the end

File: docs/tables/aws_elastic_beanstalk_application.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Table: aws_elastic_beanstalk_application
+
+AWS Elastic Beanstalk is an easy-to-use service. It offers the fastest and simplest way to deploy developer's applications without worrying about the infrastructure while maintaining high availability.
```
Comment (Paulami30) [inline docs/tables/aws_elastic_beanstalk_application.md:3]:
```suggestion
AWS Elastic Beanstalk offers the fastest and simplest way to deploy developer's applications without worrying about the infrastructure while maintaining high availability.
```

File: docs/tables/aws_elastic_beanstalk_application.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  aws_elastic_beanstalk_application;
+```
+
+
+### Info of resource life cycle configuration for each application
+
+```sql
+select
+  name,
+  jsonb_pretty(resource_lifecycle_config)
```
Comment (Paulami30) [inline docs/tables/aws_elastic_beanstalk_application.md:27]: expand the json object

File: aws/table_aws_elastic_beanstalk_application.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+	data, err := svc.DescribeApplications(params)
+
+	if len(data.Applications) > 0 && data.Applications[0] != nil {
+		return data.Applications[0], nil
+	}
+
+	return nil, nil
+}
+
+func getAwsElasticBeanstalkApplicationTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_elastic_beanstalk_application.go:169]:
```suggestion
func listAwsElasticBeanstalkApplicationTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: aws-test/tests/aws_elastic_beanstalk_application/test-get-expected.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+[
+    {
+      "arn": "{{ output.resource_aka.value }}",
+      "name": "{{ resourceName }}",
+      "tags": {
+        "name": "{{ resourceName }}"
+       }
+    }
```
Comment (khushboo9024) [inline aws-test/tests/aws_elastic_beanstalk_application/test-get-expected.json:8]: Formatting is not correct, same in all the below expected.json.

File: aws-test/tests/aws_elastic_beanstalk_application/test-get-query.sql [hunk 6]
```diff
@@ -0,0 +1,3 @@
+select arn , name , tags
```
Comment (khushboo9024) [inline aws-test/tests/aws_elastic_beanstalk_application/test-get-query.sql:1]:
```suggestion
select arn, name, tags
```
Remove extra spaces from all the below queries as well.

File: aws-test/tests/aws_elastic_beanstalk_application/variables.tf [hunk 7]
```diff
@@ -0,0 +1,84 @@
+
+variable "resource_name" {
+  type        = string
+  default     = "turbot-test-20200125-create-update"
+  description = "Name of the resource used throughout the test."
+}
+
+variable "aws_profile" {
+  type        = string
+  default     = "integration-tests"
```
Comment (khushboo9024) [inline aws-test/tests/aws_elastic_beanstalk_application/variables.tf:10]:
```suggestion
  default     = "default"
```

File: aws/table_aws_elastic_beanstalk_application.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+		GetMatrixItem: BuildRegionList,
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "name",
+				Description: "The name of the application.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ApplicationName"),
+			},
+			{
+				Name:        "arn",
```
Comment (khushboo9024) [inline aws/table_aws_elastic_beanstalk_application.go:35]: keep the column name as `application_arn`

File: aws/table_aws_elastic_beanstalk_application.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+				Transform:   transform.FromField("ResourceTags").Transform(getElasticBeanstalkApplicationTurbotTags),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listAwsElasticBeanstalkApplications(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	// TODO put me in helper function
```
Comment (khushboo9024) [inline aws/table_aws_elastic_beanstalk_application.go:105]:
```suggestion
```

File: aws/table_aws_elastic_beanstalk_application.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	return nil, nil
+}
+
+//// HYDRATE FUNCTIONS
+
+func getAwsElasticBeanstalkApplication(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	logger := plugin.Logger(ctx)
+	logger.Trace("getAwsElasticBeanstalkApplication")
+
+	// TODO put me in helper function
```
Comment (khushboo9024) [inline aws/table_aws_elastic_beanstalk_application.go:140]:
```suggestion
```

File: docs/tables/aws_elastic_beanstalk_application.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  description,
+  date_created,
+  date_updated,
+  versions
+from
+  aws_elastic_beanstalk_application;
+```
+
+
+### Info of resource life cycle configuration for each application
```
Comment (khushboo9024) [inline docs/tables/aws_elastic_beanstalk_application.md:22]:
```suggestion
### Get resource life cycle configuration details for each application
```

File: aws/table_aws_elastic_beanstalk_application.go [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	return nil, nil
+}
+
+func listAwsElasticBeanstalkApplicationTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	logger := plugin.Logger(ctx)
+	logger.Trace("listAwsElasticBeanstalkApplicationTags")
+
+	// TODO put me in helper function
```
Comment (khushboo9024) [inline aws/table_aws_elastic_beanstalk_application.go:173]:
```suggestion
```

File: aws/table_aws_elastic_beanstalk_application.go [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+		GetMatrixItem: BuildRegionList,
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "name",
+				Description: "The name of the application.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ApplicationName"),
+			},
+			{
+				Name:        "application_arn",
```
Comment (Paulami30) [inline aws/table_aws_elastic_beanstalk_application.go:35]:
```suggestion
				Name:        "arn",
```

File: aws/table_aws_elastic_beanstalk_application.go [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+	// Build the params
+	params := &elasticbeanstalk.DescribeApplicationsInput{
+		ApplicationNames: []*string{aws.String(name)},
+	}
+
+	// Get call
+	data, err := svc.DescribeApplications(params)
+
+	if len(data.Applications) > 0 && data.Applications[0] != nil {
```
Comment (Paulami30) [inline aws/table_aws_elastic_beanstalk_application.go:160]:
```suggestion
	if data.Applications[0] != nil  && len(data.Applications) > 0 {
```

File: aws/plugin.go [hunk 15]
```diff
@@ -69,7 +69,17 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_ec2_transit_gateway":                tableAwsEc2TransitGateway(ctx),
 			"aws_ec2_transit_gateway_route_table":    tableAwsEc2TransitGatewayRouteTable(ctx),
 			"aws_ec2_transit_gateway_vpc_attachment": tableAwsEc2TransitGatewayVpcAttachment(ctx),
+			"aws_ecs_cluster":                        tableAwsEcsCluster(ctx),
+    	"aws_ecs_task_definition":                tableAwsEcsTaskDefinition(ctx),
```
Comment (Paulami30) [inline aws/plugin.go]: fix indentation

**Review by Paulami30 — APPROVED** [review]

**Comment by Paulami30** [comment]
LGTM

**Comment by Paulami30** [comment]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**